src/v/coproc/tests/coproc_test_fixture.cc

```diff
@@ -98,36 +78,64 @@ ss::future<coproc_test_fixture::opt_reader_data_t> coproc_test_fixture::drain(
     });
 }
 
+void interpret_batch(
+  std::size_t& n_records_read,
+  model::offset& offset,
+  model::record_batch_reader::data_t& data,
+  model::record_batch_reader::data_t&& batch) {
+    if (batch.empty()) {
+        return;
+    }
+    offset = batch.back().last_offset();
+    for (model::record_batch& r : batch) {
+        n_records_read += r.record_count();
+        data.push_back(std::move(r));
+    }
+}
+
+model::offset next_offset(model::offset o) {
+    if (o == model::model_limits<model::offset>::min()) {
+        return model::offset(0);
+    }
+    return o + model::offset(1);
+}
+
 ss::future<model::record_batch_reader::data_t> coproc_test_fixture::do_drain(
   kafka::partition_wrapper pw,
   std::size_t limit,
   model::timeout_clock::time_point timeout) {
     return ss::do_with(
+      model::timeout_clock::time_point(),
       static_cast<std::size_t>(0),
-      model::offset(0),
+      model::model_limits<model::offset>::min(),
       model::record_batch_reader::data_t(),
-      [timeout, limit, pw](auto& acc, auto& offset, auto& b) mutable {
-          // Loop until the expected number of records is read
-          // OR a non-configurable timeout has been reached
-          return poll_until(
-                   [&b, &acc, &offset, &timeout, limit, pw]() mutable {
-                       return pw.make_reader(log_rdr_cfg(model::offset(acc)))
-                         .then([&timeout](model::record_batch_reader reader) {
+      [limit, timeout, pw](
+        model::timeout_clock::time_point now,
+        std::size_t& n_records_read,
+        model::offset& offset,
+        model::record_batch_reader::data_t& data) mutable {
+          return ss::do_until(
+                   [&n_records_read, &now, limit, timeout] {
+                       now = model::timeout_clock::now();
+                       return (n_records_read >= limit) || (now > timeout);
+                   },
+                   [&n_records_read, &offset, &data, pw]() mutable {
+                       if (pw.last_stable_offset() == offset) {
+                           return ss::sleep(20ms);
```

Comment:
will this still work if instead of sleeping here (as what i assume is a back-off heuristic), you sleep below when the record reader returns no records?

Reply:
I just tested that, yes it works as well, maybe that is a tiny bit better , since we can avoid querying log for its offsets?

Reply:
its less about performance and more about clean semantics